local-cluster/tests/local_cluster.rs

```diff
@@ -406,15 +391,141 @@ fn test_kill_partition() {
+    // 5) runs `on_partition_resolved`
+    let mut leader_schedule = vec![];
+    let num_slots_per_validator = 8;
+    let partitions: [&[usize]; 3] = [
+        &[(failures_stake as usize)],
+        &[(alive_stake_1 as usize)],
+        &[(alive_stake_2 as usize)],
+    ];
+    let validator_keys: Vec<_> = iter::repeat_with(|| Arc::new(Keypair::new()))
+        .take(partitions.len())
+        .collect();
+    for (i, k) in validator_keys.iter().enumerate() {
+        let num_slots = {
+            if i == 0 {
+                // Set up the leader to have 50% of the slots
+                num_slots_per_validator * (partitions.len() - 1)
+            } else {
+                num_slots_per_validator
+            }
+        };
+        for _ in 0..num_slots {
+            leader_schedule.push(k.pubkey())
+        }
+    }
+    info!("leader_schedule: {}", leader_schedule.len());
+
+    let validator_to_kill = validator_keys[0].pubkey();
+    let on_partition_start = |cluster: &mut LocalCluster| {
+        info!("Killing validator with id: {}", validator_to_kill);
+        cluster.exit_node(&validator_to_kill);
+    };
+    run_cluster_partition(
+        &partitions,
+        Some((
+            LeaderSchedule::new_from_schedule(leader_schedule),
+            validator_keys,
+        )),
+        on_partition_start,
+        on_partition_resolved,
+    )
+}
+
+#[test]
+#[serial]
+fn test_kill_partition_switch_threshold_no_progress() {
+    let max_switch_threshold_failure_pct = 1.0 - 2.0 * SWITCH_FORK_THRESHOLD;
+    let total_stake = 10_000;
+    let max_failures_stake = (max_switch_threshold_failure_pct * total_stake as f64) as u64;
+
+    let failures_stake = max_failures_stake;
+    let total_alive_stake = total_stake - failures_stake;
+    let alive_stake_1 = total_alive_stake / 2;
+    let alive_stake_2 = total_alive_stake - alive_stake_1;
+
+    // Check that no new roots were set 400 slots after partition resolves (gives time
+    // for lockouts built during partition to resolve and gives validators an opportunity
+    // to try and switch forks)
+    let on_partition_resolved = |cluster: &mut LocalCluster| {
+        cluster.check_no_new_roots(400, &"PARTITION_TEST");
+    };
+
+    // This kills `max_failures_stake`, so no progress should be made
+    run_kill_partition_switch_threshold(
+        failures_stake,
+        alive_stake_1,
+        alive_stake_2,
+        on_partition_resolved,
+    );
+}
+
+#[test]
+#[serial]
+fn test_kill_partition_switch_threshold() {
+    let max_switch_threshold_failure_pct = 1.0 - 2.0 * SWITCH_FORK_THRESHOLD;
+    let total_stake = 10_000;
+    let max_failures_stake = (max_switch_threshold_failure_pct * total_stake as f64) as u64;
+    let failures_stake = max_failures_stake - 2;
+    let total_alive_stake = total_stake - failures_stake;
+    let alive_stake_1 = total_alive_stake / 2;
+    let alive_stake_2 = total_alive_stake - alive_stake_1;
```

Comment:
@aeyakovenko I think this test highlights a potential liveness problem caused by the interaction of the fork choice rule and the switching threshold (you can adjust the `max_failures` stake and the distribution of `alive_stake_1` and `alive_stake_2` to test the below example). 

Recall the current behavior when a validator on fork `A` detects a heavier fork `B`, the validator will 

1) Stop voting on its own fork `A`
2) Continue generating blocks on its own fork `A`
3) Wait for a switching proof to fork `B`, at which point it will vote and reset to `B`


Now consider the below example, where the switching threshold is `38%` (need >`38%` to switch). In the example, a partition occurs such that:

1) One partition has `<=38%` of the stake but is heavier (it has a lucky run of slots)
2) One partition has `>62%` but is lighter

``` 
             4- 5- 6- 7-8-9-10  (38%)
            / 
1----2
           \
             3-11 (62%)
```
In the above example there was a brief run where the `38%` partition got more blocks.

Then when the partitions resolve,  partition `2` will stop voting and wait for a switching proof to the heavier partition `1` (code here: https://github.com/solana-labs/solana/blob/master/core/src/replay_stage.rs#L1451-L1459) but `1` will never have enough stake for a switching proof). 

Although `2` will keep generating more blocks on it's own fork because it has more stake, it's not voting, so the weight doesn't catch up (original design was based on fact that fork with more stake/slots would eventually catch up due to generating more blocks, but this reasoning seems flawed) because fork `1` is both generating blocks and voting.

Reply:
yea, i see.  What's weird is that 38% fork can construct a switching proof to the 62% fork, but its lighter

Reply:
i think the accumulation of weight ignores the expiration of votes.  that's part of the problem.  because eventually the 38% fork would expire all their votes and the 62 will catch up.